api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+	// This field is mandatory.
+	Action AdminNetPolRuleAction `json:"action"`
+
+	// Ports allows for matching on traffic based on port and protocols.
+	// This field is mandatory.
+	Ports AdminNetworkPolicyPorts `json:"ports"`
+
+	// List of sources from which traffic will be allowed/denied/passed to the entities
+	// selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR
+	// operation. If this field is empty, this rule matches no sources.
+	// If this field is present and contains at least one item, this rule
+	// allows/denies/passes traffic from the defined AdminNetworkPolicyPeer(s)
+	// If it is empty no traffic is matched by the AdminNetworkPolicyIngressRule.
+	From []AdminNetworkPolicyPeer `json:"from"`
+}
+
+// AdminNetworkPolicyEgressRule describes an action to take on a particular
+// set of traffic originating from pods selected by a AdminNetworkPolicy's
+// Subject field. The traffic must match both ports and to.
+type AdminNetworkPolicyEgressRule struct {
+	// Name is an identifier for this rule, that should be no more than 100 characters
+	// in length.
+	// +optional
+	// +kubebuilder:validation:MaxLength=100
+	Name string `json:"name,omitempty"`
+
+	// Action specifies whether this rule must pass, allow or deny traffic.
+	// Allow: allows the selected traffic
+	// Deny: denies the selected traffic
+	// Pass: instructs the selected traffic to skip any remaining positive priority
+	// ANP rules, and then pass execution to any NetworkPolicies that select the pod.
+	// If the pod is not selected by any NetworkPolicies then execution
+	// is passed to any Priority=0 ANP rules that select the pod, whereupon the traffic
+	// is allowed if not dropped by any Priority=0 ANP rules.
+	// This field is mandatory.
+	Action AdminNetPolRuleAction `json:"action"`
+
+	// Ports allows for matching on traffic based on port and protocols.
+	// This field is mandatory.
+	Ports AdminNetworkPolicyPorts `json:"ports"`
+
+	// List of destinations to which traffic will be allowed/denied/passed from the entities
+	// selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR
+	// operation. If this field is empty, this rule matches no destinations.
+	// If this field is present and contains at least one item, this rule
+	// allows/denies/passes traffic to the defined AdminNetworkPolicyPeer(s)
+	// If it is empty no traffic is matched by the AdminNetworkPolicyEgressRule.
+	To []AdminNetworkPolicyPeer `json:"to"`
+}
+
+// AdminNetworkPolicyPorts handles selection of traffic based on port(s).
+// Exactly one of the fields must be defined.
+type AdminNetworkPolicyPorts struct {
+	// AllPorts cannot be "false" when it is set
+	// AllPorts allows the user to select all ports for all protocols, thus not
+	// selecting traffic based on L4 principles.
+	// If "true" then all ports are selected for all protocols.
+	// +optional
+	AllPorts *bool `json:"allPorts,omitempty"`
```

Comment:
all parts of this are terrible :stuck_out_tongue_closed_eyes: 

Why is it better to force people to say

    - action: Allow
      from: ...
      ports:
        allPorts: true

rather than just

    - action: Allow
      from: ...

It seems pretty clear to me that the second rule implies "does not filter based on ports", in the same way that it implies "does not filter based on uid" and "does not filter based on time of day". (Indeed, if we are going to say that absence-of-`ports` _can't_ imply "does not filter based on ports", then that seems to imply that we _can't_ add new things to possibly filter on in the future, since we can't retroactively require people to add `timeOfDay: { anyTimeOfDay: true }` to their policies, and we aren't willing to let `nil` imply "not relevant".)

If we are going to have a required thing, then having a boolean field that can only be `true` seems wrong

Reply:
@thockin

Reply:
I tend to agree with Dan and Rahul here, it's all about finding that balance of implicitness vs explicitness